File: BOJ/2164.cpp
#include <stdio.h>
#include <algorithm>
#include <queue>
using namespace std;
int arr[500001];
int n, cnt;
bool flag;
queue <int> q;
int main() {
	scanf("%d", &n);
	for (int i = 1; i <= n; i++) {
		arr[i] = i;
		q.push(i);
	}
	while (cnt != n) {
		if (!flag){
			if (cnt == n - 1) printf("%d\n", q.front());
			q.pop();
			flag = true;
			cnt++;
		}
		else {
			int front = q.front();
			q.push(front);
			q.pop();
			flag = false;
		}
	}
	return 0;
}

File: BOJ(Java 공부용)/2164.java
import java.io.*;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int[] arr = new int[n + 1];
        ArrayDeque<Integer> q = new ArrayDeque<>();
        for (int i = 1; i <= n; i++) {
            arr[i] = i;
            q.add(i);
        }
        int cnt = 0;
        boolean flag = false;
        while (cnt != n) {
            if (!flag) {
                if (cnt == n - 1) {
                    System.out.println(q.peek());
                    return;
                }
                q.poll();
                flag = true;
                cnt++;
            } else {
                int front = q.peek();
                q.add(front);
                q.poll();
                flag = false;
            }
        }
    }
}